File: frontend/src/utils/agentReplyCleanup.js
/**
 * 에이전트 답변 후처리 (스트림 완료 후): TR 목록 이중 표기·내부 안내 문구 제거.
 * 백엔드 api/agent_reply_cleanup.py 와 동작을 맞춤.
 */

const LEGACY_HINT =
  /아래는 list_transports 조회 결과입니다\.[\s\S]*?(?=\s*[-*]?\s*[A-Z]{2,4}\d{6,}\s*:)/

const TR_LINE_RE = /^\s*(?:[-*]\s+)?[A-Z]{2,4}\d{6,}\s*:/

function trIdFromLine(line) {
  const m = line.match(/^\s*(?:[-*]\s+)?([A-Z]{2,4}\d{6,})\s*:/)
  return m ? m[1] : null
}

function isTrLine(line) {
  return TR_LINE_RE.test(line)
}

function trIdsFromLines(lines) {
  return lines.map(trIdFromLine).filter(Boolean)
}

function idsEqual(a, b) {
  return a.length === b.length && a.every((v, k) => v === b[k])
}

function normalizeGluedTrHeading(text) {
  return text.replace(/([^\s\n])(###\s*TR\s*목록)/gi, '$1\n\n$2')
}

/**
 * 문자열-aware로 첫 JSON 값 `{...}` 또는 `[...]`의 끝 인덱스까지 파싱 (json.JSONDecoder.raw_decode 대응).
 * @returns {{ value: unknown, end: number } | null}
 */
function rawDecodeFirstJsonValue(s) {
  let i = 0
  while (i < s.length && /\s/.test(s[i])) i += 1
  if (i >= s.length || (s[i] !== '{' && s[i] !== '[')) return null
  const start = i
  const stack = []
  let inStr = false
  let esc = false
  for (; i < s.length; i += 1) {
    const c = s[i]
    if (esc) {
      esc = false
      continue
    }
    if (inStr) {
      if (c === '\\') esc = true
      else if (c === '"') inStr = false
      continue
    }
    if (c === '"') {
      inStr = true
      continue
    }
    if (c === '{') stack.push('}')
    else if (c === '[') stack.push(']')
    else if (c === '}' || c === ']') {
      const expected = stack.pop()
      if (expected !== c) return null
      if (stack.length === 0) {
        const chunk = s.slice(start, i + 1)
        try {
          return { value: JSON.parse(chunk), end: i + 1 }
        } catch {
          return null
        }
      }
    }
  }
  return null
}

/** Docs MCP search 등 `{"results":[...]}` 가 답변 앞에 붙은 경우 제거 */
function stripLeadingSearchResultsJsonBlob(text) {
  if (!text || !text.includes('results')) return text
  const s = text.replace(/^\uFEFF/, '').trimStart()
  const parsed = rawDecodeFirstJsonValue(s)
  if (!parsed) return text
  const { value, end } = parsed
  if (!value || typeof value !== 'object' || !Array.isArray(value.results)) return text
  const { results } = value
  if (results.length > 0) {
    const sample = results[0]
    if (!sample || typeof sample !== 'object') return text
    const keys = ['title', 'url', 'snippet', 'id', 'topic', 'library_id']
    if (!keys.some((k) => Object.prototype.hasOwnProperty.call(sample, k))) return text
  }
  return s.slice(end).replace(/^\s+/, '')
}

function stripLegacyListTransportsHint(text) {
  if (!text || !text.includes('아래는 list_transports')) return text
  return text.replace(LEGACY_HINT, '')
}

function dedupeByTrailingTrLines(beforeRaw, after, afterTr) {
  const beforeLines = beforeRaw.split('\n')
  const trTail = []
  let cut = beforeLines.length
  let i = beforeLines.length - 1
  while (i >= 0) {
    const line = beforeLines[i]
    if (line.trim() === '') {
      if (trTail.length) break
      cut = i
      i -= 1
      continue
    }
    if (isTrLine(line)) {
      trTail.unshift(line)
      cut = i
      i -= 1
      continue
    }
    break
  }
  if (!trTail.length || !idsEqual(trIdsFromLines(trTail), trIdsFromLines(afterTr))) return null
  const newBefore = beforeLines.slice(0, cut).join('\n').replace(/\s+$/, '')
  const sep = newBefore ? '\n\n' : ''
  return `${newBefore}${sep}${after}`
}

function allLinesMatchTr(lines) {
  const nonEmpty = lines.filter((ln) => ln.trim())
  return nonEmpty.length > 0 && nonEmpty.every((ln) => TR_LINE_RE.test(ln))
}

function dedupeByParagraph(beforeRaw, after, afterIds) {
  const t = beforeRaw.trim()
  if (!t) return null
  const paras = t.split(/\n\s*\n+/)
  for (let pi = paras.length - 1; pi >= 0; pi -= 1) {
    const p = paras[pi].trim()
    if (!p) continue
    const lines = p.split('\n')
    if (!allLinesMatchTr(lines)) continue
    if (!idsEqual(trIdsFromLines(lines), afterIds)) continue
    const newParas = paras.filter((_, i) => i !== pi)
    const newBefore = newParas.map((x) => x.trim()).filter(Boolean).join('\n\n').replace(/\s+$/, '')
    const sep = newBefore ? '\n\n' : ''
    return `${newBefore}${sep}${after}`
  }
  return null
}

function dedupeBySlidingWindow(beforeRaw, after, afterTr) {
  const beforeLines = beforeRaw.split('\n')
  const n = afterTr.length
  const afterIds = trIdsFromLines(afterTr)
  if (n === 0 || beforeLines.length < n) return null
  for (let s = 0; s <= beforeLines.length - n; s += 1) {
    const chunk = beforeLines.slice(s, s + n)
    if (!chunk.every((ln) => isTrLine(ln))) continue
    if (!idsEqual(trIdsFromLines(chunk), afterIds)) continue
    const newLines = [...beforeLines.slice(0, s), ...beforeLines.slice(s + n)]
    const newBefore = newLines.join('\n').replace(/\s+$/, '')
    const sep = newBefore ? '\n\n' : ''
    return `${newBefore}${sep}${after}`
  }
  return null
}

function dedupeTrBlockBeforeTrHeading(text) {
  const m = text.match(/###\s*TR\s*목록/i)
  if (!m || m.index === undefined) return text
  const idx = m.index
  const beforeRaw = text.slice(0, idx)
  let after = text.slice(idx).trimStart()

  const afterLines = after.split('\n')
  const afterTr = []
  let j = 1
  while (j < afterLines.length) {
    const line = afterLines[j]
    if (isTrLine(line)) {
      afterTr.push(line)
      j += 1
      continue
    }
    if (line.trim() === '' && afterTr.length) break
    if (afterTr.length) break
    j += 1
  }

  if (!afterTr.length) return text
  const afterIds = trIdsFromLines(afterTr)

  return (
    dedupeByTrailingTrLines(beforeRaw, after, afterTr) ||
    dedupeByParagraph(beforeRaw, after, afterIds) ||
    dedupeBySlidingWindow(beforeRaw, after, afterTr) ||
    text
  )
}

/**
 * @param {string} text
 * @returns {string}
 */
export function cleanAgentReplyText(text) {
  if (!text) return text
  let t = stripLeadingSearchResultsJsonBlob(text)
  t = normalizeGluedTrHeading(t)
  t = stripLegacyListTransportsHint(t)
  t = dedupeTrBlockBeforeTrHeading(t)
  return t.trim()
}

File: backend/api/agent_reply_cleanup.py
"""
에이전트 최종 답변 후처리: TR 목록 중복·내부 안내 문구 유출 제거.
"""
from __future__ import annotations

import json
import re

_HEADING_TR = re.compile(r"###\s*TR\s*목록", re.IGNORECASE)
# 과거 list_transports에 넣었던 사용자 노출용 안내(첫 TR 라인 직전까지)
_LEGACY_HINT = re.compile(
    r"아래는 list_transports 조회 결과입니다\.[\s\S]*?(?=\s*[-*]?\s*[A-Z]{2,4}\d{6,}\s*:)",
)


def strip_legacy_list_transports_hint(text: str) -> str:
    """모델이 도구 안내 문구를 답변에 그대로 붙인 경우 제거."""
    if not text or "아래는 list_transports" not in text:
        return text
    return _LEGACY_HINT.sub("", text, count=1)


def _tr_ids_from_line_block(lines: list[str]) -> list[str]:
    ids: list[str] = []
    for line in lines:
        m = re.match(r"^\s*(?:[-*]\s+)?([A-Z]{2,4}\d{6,})\s*:", line)
        if m:
            ids.append(m.group(1))
    return ids


_TR_LINE_RE = re.compile(r"^\s*(?:[-*]\s+)?[A-Z]{2,4}\d{6,}\s*:")


def _all_lines_match_tr(lines: list[str]) -> bool:
    non_empty = [ln for ln in lines if ln.strip()]
    return bool(non_empty) and all(_TR_LINE_RE.match(ln) for ln in non_empty)


def _dedupe_by_trailing_tr_lines(before_raw: str, after: str, after_tr: list[str]) -> str | None:
    """`before` 끝이 TR 줄들로만 끝나면 그 블록 제거. 실패 시 None."""
    before_lines = before_raw.split("\n")
    tr_tail: list[str] = []
    cut = len(before_lines)
    i = len(before_lines) - 1
    while i >= 0:
        line = before_lines[i]
        if line.strip() == "":
            if tr_tail:
                break
            cut = i
            i -= 1
            continue
        if _TR_LINE_RE.match(line):
            tr_tail.insert(0, line)
            cut = i
            i -= 1
            continue
        break
    if not tr_tail or _tr_ids_from_line_block(tr_tail) != _tr_ids_from_line_block(
        after_tr
    ):
        return None
    new_before = "\n".join(before_lines[:cut]).rstrip()
    sep = "\n\n" if new_before else ""
    return f"{new_before}{sep}{after}"


def _dedupe_by_tr_paragraph(before_raw: str, after: str, after_ids: list[str]) -> str | None:
    """
    빈 줄로 구분된 문단 중, 전 줄이 TR 형식이고 TR id 순서가 헤딩 아래와 동일한 문단을
    뒤에서부터 찾아 제거 (티켓/종속성 문단이 그 뒤에 있을 때).
    """
    if not before_raw.strip():
        return None
    paras = re.split(r"(?:\r?\n\s*){2,}", before_raw.strip())
    if len(paras) < 1:
        return None
    for pi in range(len(paras) - 1, -1, -1):
        p = paras[pi].strip()
        if not p:
            continue
        lines = p.split("\n")
        if not _all_lines_match_tr(lines):
            continue
        if _tr_ids_from_line_block(lines) != after_ids:
            continue
        new_paras = [x for i, x in enumerate(paras) if i != pi]
        new_before = "\n\n".join(x.strip() for x in new_paras if x.strip()).rstrip()
        sep = "\n\n" if new_before else ""
        return f"{new_before}{sep}{after}"
    return None


def _dedupe_by_sliding_window(before_raw: str, after: str, after_tr: list[str]) -> str | None:
    """연속 n줄이 모두 TR이고 id 순서가 헤딩 아래와 같으면 해당 구간 삭제 (문단 구분 없을 때)."""
    before_lines = before_raw.split("\n")
    n = len(after_tr)
    if n == 0 or len(before_lines) < n:
        return None
    after_ids = _tr_ids_from_line_block(after_tr)
    for s in range(0, len(before_lines) - n + 1):
        chunk = before_lines[s : s + n]
        if not all(_TR_LINE_RE.match(ln) for ln in chunk):
            continue
        if _tr_ids_from_line_block(chunk) != after_ids:
            continue
        new_lines = before_lines[:s] + before_lines[s + n :]
        new_before = "\n".join(new_lines).rstrip()
        sep = "\n\n" if new_before else ""
        return f"{new_before}{sep}{after}"
    return None


def dedupe_tr_block_before_tr_heading(text: str) -> str:
    """
    '### TR 목록' 아래 TR 목록과 동일한 TR 블록이 헤딩 앞에 있으면 한쪽 제거.
    """
    if not text or "###" not in text:
        return text
    m = _HEADING_TR.search(text)
    if not m:
        return text
    idx = m.start()
    before_raw = text[:idx]
    after = text[idx:].lstrip()

    after_lines = after.split("\n")
    if not after_lines:
        return text
    after_tr: list[str] = []
    j = 1
    while j < len(after_lines):
        line = after_lines[j]
        if _TR_LINE_RE.match(line):
            after_tr.append(line)
            j += 1
            continue
        if line.strip() == "" and after_tr:
            break
        if after_tr:
            break
        j += 1

    if not after_tr:
        return text

    after_ids = _tr_ids_from_line_block(after_tr)

    out = _dedupe_by_trailing_tr_lines(before_raw, after, after_tr)
    if out is not None:
        return out
    out = _dedupe_by_tr_paragraph(before_raw, after, after_ids)
    if out is not None:
        return out
    out = _dedupe_by_sliding_window(before_raw, after, after_tr)
    if out is not None:
        return out
    return text


def normalize_glued_tr_heading(text: str) -> str:
    """'없음### TR 목록' 처럼 붙은 헤딩 앞에 줄바꿈 삽입."""
    return re.sub(
        r"([^\s\n])(###\s*TR\s*목록)",
        r"\1\n\n\2",
        text,
        flags=re.IGNORECASE,
    )


def strip_leading_search_results_json_blob(text: str) -> str:
    """Docs MCP 등 `search`가 반환한 {\"results\":[...]} 원문이 답변 앞(또는 공백 뒤)에 붙은 경우 제거."""
    if not text or "results" not in text:
        return text
    s = text.lstrip()
    if not s.startswith("{"):
        return text
    dec = json.JSONDecoder()
    try:
        obj, end = dec.raw_decode(s)
    except json.JSONDecodeError:
        return text
    if not isinstance(obj, dict):
        return text
    results = obj.get("results")
    if not isinstance(results, list):
        return text
    if results:
        sample = results[0]
        if not isinstance(sample, dict):
            return text
        if not any(
            k in sample for k in ("title", "url", "snippet", "id", "topic", "library_id")
        ):
            return text
    return s[end:].lstrip()


def clean_agent_reply_text(text: str) -> str:
    """스트림/비스트림 공통 후처리."""
    if not text:
        return text
    t = strip_leading_search_results_json_blob(text)
    t = normalize_glued_tr_heading(t)
    t = strip_legacy_list_transports_hint(t)
    t = dedupe_tr_block_before_tr_heading(t)
    return t.strip()
